src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,482 @@
+                      \FROM [" <> domainTableName d <> "]\
+                      \ WHERE rowkey = ?\
+                      \ ORDER BY blockheight DESC, version DESC\
+                      \ LIMIT 1"
+                qry db stmt [SText kstr] [RBlob]
+     case result of
+         [] -> return Nothing
+         [[SBlob a]] -> return $ decode $ fromStrict a
+         err -> internalError $ "readRow: Expected (at most) a single result, but got: " <> T.pack (show err)
+
+writeSys :: (AsString k, ToJSON v) => WriteType -> Domain k v -> k -> v -> BlockHandler SQLiteEnv ()
+writeSys wt d k v = do
+        BlockVersion bh version <- gets _bsBlockVersion
+        txid <- gets _bsTxId
+        callDb (write (getKeyString k) bh version txid)
+        let toTableName (Utf8 str) = TableName $ toS str
+        record (toTableName $ domainTableName d) d k v
+  where
+    getKeyString key = case d of
+      KeySets -> convKeySetName key
+      Modules -> convModuleName key
+      Namespaces -> convNamespaceName key
+      Pacts -> convPactId key
+      UserTables _ -> error "impossible"
+    write key bh version txid db = case wt of
+        Insert -> backendWriteInsert key (domainTableName d) bh version txid v db
+        Update -> backendWriteUpdate key (domainTableName d) bh version txid v db
+        Write -> backendWriteWrite key (domainTableName d) bh version txid v db
+
+backendWriteInsert :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteInsert key tn bh version txid v db =
+  exec' db ("INSERT INTO [" <> tn <> "] ('rowkey','blockheight','version','txid','tabledata') VALUES (?,?,?,?,?);")
+    [SText key
+    , SInt (fromIntegral bh)
+    , SInt (fromIntegral version)
+    , SInt (fromIntegral txid)
+    , SBlob (toStrict (Data.Aeson.encode v))]
+
+backendWriteUpdate :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteUpdate key tn bh version txid v db =
+  exec' db ("UPDATE [" <> tn <> "] SET rowkey = ?, blockheight = ?, version =  ?, txid = ?, tabledata = ? ;")
+    [SText key
+    , SInt (fromIntegral bh)
+    , SInt (fromIntegral version)
+    , SInt (fromIntegral txid)
+    , SBlob (toStrict (Data.Aeson.encode v))]
+
+-- don't need the initial select, it will just fail. Look at pact
+backendWriteWrite :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteWrite key tn bh version txid v db =
+  exec' db ("INSERT OR REPLACE INTO [" <> tn <> "] ('rowkey','blockheight','version','txid','tabledata') VALUES (?,?,?,?,?) ;")
+    [SText key, SInt (fromIntegral bh), SInt (fromIntegral version), SInt (fromIntegral txid), SBlob (toStrict (Data.Aeson.encode v))]
+
+writeUser :: WriteType -> Domain RowKey (ObjectMap PactValue) -> RowKey -> ObjectMap PactValue -> BlockHandler SQLiteEnv ()
+writeUser wt d k row = do
+      case d of
+        (UserTables _) -> do
```

Comment:
You're calling this function based on already having `case`d on the `Domain` in `writeRow` -- just pass the table name instead

Reply:
Also shouldn't this be warning/not building (per `-Werror` on incomplete pattern match??)

Reply:
@slpopejoy, I had also thought that there should be a case warning, HOWEVER, GHC is able to figure out that there is only ONE conceivable value for the argument `d` because of the type signature of `writeUser`. Which does imply that I don't need the case statement. I'l get rid of the case statement.